### ISSUE TITLE
[Tracking Issue] Create/Convert ADLS Gen2 (data_lake) using Pipeline Architecture

### ISSUE THREAD

**Issue by thovoll** [Azure/azure-sdk-for-rust#496]
This is a tracking issue for creating/converting ADLS Gen2 (data_lake) operations using the pipeline architecture.

### Overview 

Previous to the pipeline architecture, each operation was independently controlled meaning there was no central way to control the way operations were performed. 

The pipeline architecture on the other hand allows users of the SDK to create pipelines through which outgoing requests and incoming responses are processed. The pipeline are composed of "policies" which can both change the outgoing request and the incoming response. Policies are usually either on a "per-request" basis (meaning they are called only once per operation) or "per-retry" (meaning they are applied every time an operation is attempted). 

Examples of policies are the `TelemetryPolicy` which adds various telemetry headers to the request and the retry policy which checks whether an incoming response was successful and if it detects a transient failure, retries the request. 

### Converting to the Architecture

For an example of adding the first operation using the pipeline architecture, take a look at #358 and its follow-ups #451 and #491. 
Once the basic structure is established, adding new operations is easier: #495.

No **existing** operations have been converted yet, but the basic steps are shown in the CosmosDB tracking issue: #290.

## Todos

- [ ] **Files** #610 
    - [X] [`create_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create) #358, #451, #491
    - [X] [`create_file_if_not_exists`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create) #495
    - [X] [`rename_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create) #495 
    - [X] [`rename_file_if_not_exists`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create) #506 
    - [X] [`delete_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/delete) #512
    - [ ] [`lease_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/lease)
    - [X] [`get_file_properties`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/get-properties)
    - [x] [`get_file_status`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/get-properties)
    - [x] [`get_file_access_control_list`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/get-properties)
    - [X] [`read_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/read)
    - [X] [`append_to_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update) #495
    - [X] [`flush_file`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update) #495 
    - [X] [`set_file_properties`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update)
    - [x] [`set_file_access_control_list`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update)

- [ ] **Directories** #610
    - [X] [`create_directory`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create)
    - [X] [`create_directory_if_not_exists`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create)
    - [X] [`rename_directory`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create)
    - [X] [`rename_directory_if_not_exists`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create)
    - [X] [`delete_directory`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/delete)
    - [X] [`get_directory_properties`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/get-properties)
    - [x] [`get_directory_status`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/get-properties)
    - [x] [`get_directory_access_control_list`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/get-properties)
    - [ ] [`lease_directory`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/lease)
    - [X] [`list_directory`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/list)
    - [x] [`set_directory_properties`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update)
    - [x] [`set_directory_access_control_list`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update)

- [X] **File Systems** #597 
    - [X] [`list_file_systems`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/filesystem/list)
    - [X] [`create_file_system`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/filesystem/create)
    - [X] [`delete_filesystem`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/filesystem/delete)
    - [X] [`get_file_system_properties`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/filesystem/get-properties)
    - [X] [`set_file_system_properties`](https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/filesystem/set-properties)

## Related Issues

* #372 
* #426
* #489
* #490 
* #491

**Comment by roeap** [comment]
I just opened a PR for all the file system operations (#597), maybe that should be referenced in this Issue?

**Comment by roeap** [comment]
@thovoll - with #597 and #610 merged, we should be able to check some of these boxes ;)

- Files
  - read_file
  -  get_file_properties
  -  set_file_properties
- FileSystem
  - all
- Direcrtory
  -  create_directory
  - create_directory_if_not_exists
  - rename_directory
  - rename_directory_if_not_exists
  - delete_directory
  - get_directory_properties

**Comment by cataggar** [comment]
It looks like #853 from @roeap just about completed this. The todos just mention a `lease` fn for the file and directory clients. Is that still needed to complete this? @thovoll

**Comment by roeap** [comment]
Not sure hot @thovoll views this, but I had a look at other SDKs, as well as blob over here, and lease seems to be consistently implemented using  a dedicated LeaseClient. So I think file and directory client could be considered complete in terms of exposed functionality.

**Comment by roeap** [comment]
@cataggar - AFAIK #372 and #426 are done as well.

**Comment by cataggar** [comment]
If a `LeaseClient` is needed, please add a separate issue.
